src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,754 @@
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptj, &k[i]);
+        secp256k1_ge_set_gej(&nonce_pt[i], &nonce_ptj);
+        secp256k1_declassify(ctx, &nonce_pt[i], sizeof(nonce_pt));
+        secp256k1_scalar_clear(&k[i]);
+    }
+    /* nonce_pt won't be infinity because k != 0 with overwhelming probability */
+    secp256k1_musig_pubnonce_save(pubnonce, nonce_pt);
+    return ret;
+}
+
+static int secp256k1_musig_sum_nonces(const secp256k1_context* ctx, secp256k1_gej *summed_nonces, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    size_t i;
+    int j;
+
+    secp256k1_gej_set_infinity(&summed_nonces[0]);
+    secp256k1_gej_set_infinity(&summed_nonces[1]);
+
+    for (i = 0; i < n_pubnonces; i++) {
+        secp256k1_ge nonce_pt[2];
+        if (!secp256k1_musig_pubnonce_load(ctx, nonce_pt, pubnonces[i])) {
+            return 0;
+        }
+        for (j = 0; j < 2; j++) {
+            secp256k1_gej_add_ge_var(&summed_nonces[j], &summed_nonces[j], &nonce_pt[j], NULL);
+        }
+    }
+    return 1;
+}
+
+int secp256k1_musig_nonce_agg(const secp256k1_context* ctx, secp256k1_musig_aggnonce  *aggnonce, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    secp256k1_gej aggnonce_ptj[2];
+    secp256k1_ge aggnonce_pt[2];
+    int i;
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(aggnonce != NULL);
+    ARG_CHECK(pubnonces != NULL);
+    ARG_CHECK(n_pubnonces > 0);
+
+    if (!secp256k1_musig_sum_nonces(ctx, aggnonce_ptj, pubnonces, n_pubnonces)) {
+        return 0;
+    }
+    for (i = 0; i < 2; i++) {
+        if (secp256k1_gej_is_infinity(&aggnonce_ptj[i])) {
+            /* There must be at least one dishonest signer. If we would return 0
+               here, we will never be able to determine who it is. Therefore, we
+               should continue such that the culprit is revealed when collecting
+               and verifying partial signatures.
+
+               However, dealing with the point at infinity (loading,
+               de-/serializing) would require a lot of extra code complexity.
+               Instead, we set the aggregate nonce to some arbitrary point (the
+               generator). This is secure, because it only restricts the
+               abilities of the attacker: an attacker that forces the sum of
+               nonces to be infinity by sending some maliciously generated nonce
+               pairs can be turned into an attacker that forces the sum to be
+               the generator (by simply adding the generator to one of the
+               malicious nonces), and this does not change the winning condition
+               of the EUF-CMA game. */
+            aggnonce_pt[i] = secp256k1_ge_const_g;
+        } else {
+            secp256k1_ge_set_gej(&aggnonce_pt[i], &aggnonce_ptj[i]);
+        }
+    }
+    secp256k1_musig_aggnonce_save(aggnonce, aggnonce_pt);
+    return 1;
+}
+
+/* tagged_hash(aggnonce[0], aggnonce[1], agg_pk, msg) */
```

Comment:
* Should we change this to tagged_hash(agg_pk, msg, aggnonce[0], aggnonce[1]) because the first two arguments may be clear upfront, so you could precompute?
 * Should the tag include the number of nonces? Something like "MuSig2-2/noncecoef"

Sorry for bringing this up now. Please don't kill me now. :D 

We could postpone this discussion but since this is relevant for interop, it may be better to change it now than later. (As opposed to the other nonce hash which is totally not relevant.)

Reply:
From an interop perspective, wouldn't `tagged_hash(agg_pk, aggnonce[0], aggnonce[1], msg)` be the most likely to work out of the box, as it matches the equation from the paper?

In practice different choices here optimize for different users: Some users may share nonces before pk and msg are known, others may know pk and msg before they generate nonces.

Reply:
- I don't think it's necessary for the tag to include the number of nonces explicitly. "MuSig2" can implicitly mean 2 nonces and if you want to use 4 then use tag "MuSig2-4".
- I agree that different choices optimize for different users. The order I picked just mimics the order of the signature challenge. I don't think the order in the paper should be relevant because it's entirely arbitrary and the paper isn't a spec. One could argue that agg_pk, message first is more "standard". On the other hand users who are really in a hurry are more likely to pre-share nonces and therefore prefer nonces first.

Reply:
> I don't think it's necessary for the tag to include the number of nonces explicitly. "MuSig2" can implicitly mean 2 nonces and if you want to use 4 then use tag "MuSig2-4".

It's currently "MuSig/noncecoef" and not "MuSig2/noncecoef" but that really doesn't matter. Both are equally fine, I think.

> * On the other hand users who are really in a hurry are more likely to pre-share nonces and therefore prefer nonces first.

Oh indeed, I overlooked that. Then I suggest we keep it as is.